src/libraries/System.Private.CoreLib/src/System/IO/Path.cs

```diff
@@ -926,5 +926,56 @@ private static string GetRelativePath(string relativeTo, string path, StringComp
         /// Returns true if the path ends in a directory separator.
         /// </summary>
         public static bool EndsInDirectorySeparator(string path) => PathInternal.EndsInDirectorySeparator(path);
+
+        /// <summary>
+        /// Removes any redundant segments found in the specified path.
+        /// </summary>
+        /// <param name="path">A string containing an absolute or relative path that may or may not contain redundant segments.</param>
+        /// <returns><paramref name="path" /> without any redundant segments.</returns>
+        public static string RemoveRedundantSegments(string path)
+        {
+            if (path == null)
+                throw new ArgumentNullException(nameof(path));
+
+            if (PathInternal.IsEffectivelyEmpty(path))
+                throw new ArgumentException(SR.Arg_PathEmpty, nameof(path));
+
+            return RemoveRedundantSegments(path.AsSpan());
+        }
+
+        /// <summary>
+        /// Removes any redundant segments found in the specified path.
+        /// </summary>
+        /// <param name="path">A read-only span containing an absolute or relative path that may or may not contain redundant segments.</param>
+        /// <returns><paramref name="path" /> without any redundant segments.</returns>
+        public static string RemoveRedundantSegments(ReadOnlySpan<char> path)
```

Comment:
@JeremyKuhne 
Since we have two `TryRemoveRedundantSegments` methods that we don't expect to throw, then how sensitive do we want the `RemoveRedundantSegments` methods to be? I'm thinking we could throw when:

- `path` is null or effectively empty (as it is right now). Jan's comment below offers a different opinion.
- the internal call to `TryRemoveRedundantSegments` returns `false`.
- `bytesWritten` is zero.
- `path.Length` is larger than `PathInternal.MaxShortPath`, although this value is not shared by Unix (is Windows only). Ahson's comment below offers an arbitrary number 256.

Reply:
There should be no limit for path length. I also don't think we should throw if we remove everything, just return empty. `"."` shouldn't throw, it should return `""`.

I don't think we should throw for null or empty either. Just return as I described below.